WIP: Accelerate StringLatin1.indexof (string search version) Intrinsic on Power.

IBMJimmyk opened the pull request:
There are different versions of `StringLatin1.indexOf`. This work item is about `java/lang/StringLatin1.indexOf([BI[BII)I` which is used to search for the index of a String inside of another String.

IBMJimmyk commented:
I ran `StringIndexOfStringBench` from Bumblebench for performance testing:
https://github.com/adoptium/bumblebench/blob/4ec780b716e7617094cce740d3b085ca4637ee0f/net/adoptopenjdk/bumblebench/string/StringIndexOfStringBench.java

It has three options.
`stringLength` is the length of the sub string to search for.
```
Short = 8-20 characters
Medium = 80-120 characters
Long = 800-1200 characters
Ultra Long = 9000-11000 characters
```

`indexLevel` is the index in the longer string the sub string will be found at.
```
Low = 1-50
Medium = 80-120
High = 800-1200
Ultra High = 48000-51000
```

`useRandomString` is set to use a random distribution of letters for the larger string being searched in. If false, a Wikipedia excerpt is used instead that has a more typical distribution of Latin1 characters foe English text.

Results (higher number is better):
| stringLength | indexLevel | useRandomString | Baseline | Accelerated | Speedup |
| --- | --- | --- | --- | --- | --- |
| Short | Low | N | 37780388 | 65798424 | 1.74x |
| Short | Medium | N | 11950272 | 24247376 | 2.03x |
| Short | High | N | 1260837 | 2548370 | 2.02x |
| Short | Ultra High | N | 13179 | 26197 | 1.99x |
| Medium | Low | N | 37715440 | 54070372 | 1.43x |
| Medium | Medium | N | 11856987 | 22221912 | 1.87x |
| Medium | High | N | 1250761 | 2513439 | 2.01x |
| Medium | Ultra High | N | 11665 | 22283 | 1.91x |
| Long | Low | N | 11449521 | 14981757 | 1.31x |
| Long | Medium | N | 7059095 | 10581082 | 1.50x |
| Long | High | N | 1174269 | 2192076 | 1.87x |
| Long | Ultra High | N | 11500 | 22292 | 1.94x |
| Ultra Long | Low | N | 315025 | 2289270 | 7.27x |
| Ultra Long | Medium | N | 2095266 | 2185265 | 1.04x |
| Ultra Long | High | N | 256952 | 1192369 | 4.64x |
| Ultra Long | Ultra High | N | 11178 | 22100 | 1.98x |
| Short | Low | Y | 37755632 | 93064976 | 2.46x |
| Short | Medium | Y | 20522852 | 69877160 | 3.40x |
| Short | High | Y | 2193947 | 12700037 | 5.79x |
| Short | Ultra High | Y | 47860 | 241128 | 5.04x |
| Medium | Low | Y | 15013809 | 73015424 | 4.86x |
| Medium | Medium | Y | 19562160 | 56882448 | 2.91x |
| Medium | High | Y | 2187883 | 12136565 | 5.55x |
| Medium | Ultra High | Y | 47723 | 241524 | 5.06x |
| Long | Low | Y | 2705928 | 16425695 | 6.07x |
| Long | Medium | Y | 10588714 | 15586048 | 1.47x |
| Long | High | Y | 1994101 | 7730786 | 3.88x |
| Long | Ultra High | Y | 46025 | 238345 | 5.18x |
| Ultra Long | Low | Y | 307301 | 2236618 | 7.28x |
| Ultra Long | Medium | Y | 304372 | 2231100 | 7.33x |
| Ultra Long | High | Y | 270663 | 1940732 | 7.17x |
| Ultra Long | Ultra High | Y | 41331 | 217090 | 5.25x |

Performance has increased across the board. Smallest gain is 1.04x. Largest gain is 7.33x.

IBMJimmyk commented:
I have 3 new test configurations. Same as before, `StringLatin1.indexOf` is called 500 million times per test configuration.

These numbers are run on a Power 10 machine but `TR_DisableP10ArrayCmp` is set so the P10 acceleration is not used since the new intrinsic doesn't have a P10 version yet.

The data is in the attached spreadsheet but the quick summary is the new implementation is faster in all cases although some cases are almost a tie.

Test 1:
In this test the main string and the sub string are the same length and contents with one exception.
The character at the index indicated by "Mismatch ID" is different.
This tests the performance of finding a mismatch.
If the mismatch ID is not 0, only the first character of the main string matches the first character of the sub string.
If the mismatch ID is 0, a first character match of the sub string is never found.
Every configuration for string size up to 47 is tested.

Best case: Reduces time by 66%.
Worst case: Reduces time by 25%.

Test 2:
In this test the substring will match the characters in the main string starting at Match Idx.
The first character of the substring will not match any character before Match Idx.
After the first character of the substring matches at Match Idx, the rest of the substring will match as well.
The substring is determined by "Substring Length".
The main string is exactly long enough to hold the sub string at match Idx (so it is "SubstringLength+MatchIdx" long)
This tests matches at index 0-8, 14, 15, 16, 30, 31, 32, 46, 47 and 48
This tests substrings of length 1-64, 128, 256, 512, 1024

Best case: Reduces time by 81%.
Worst case: Reduces time by 3%.

Test 3:
In this test the substring will match the characters in the main string starting at Match Idx+2
The first character of the substring will match the character at Match Idx but the rest of the substring won't be found,
The first character of the substring will also match the character at Match Idx+1 but the rest of the substring still won't be found,
After the first character of the substring matches at Match Idx+2, the rest of the substring will match as well.
The substring is determined by "Substring Length".
The main string is exactly long enough to hold the sub string at match Idx+2 (so it is "SubstringLength+MatchIdx+2" long)
This tests matches at index 0-8, 14, 15, 16, 30, 31, 32, 46, 47 and 48
This tests substrings of length 1-64, 128, 256, 512, 1024

Best case: Reduces time by 74%.
Worst case: Reduces time by 2%.

[IndexOfStrPerf-2.xlsx](https://github.com/user-attachments/files/21975818/IndexOfStrPerf-2.xlsx)

IBMJimmyk commented:
I added lot of comments, cleaned up the code and squashed the commits together.

IBMJimmyk commented:
I am once again running my test that calls `StringLatin1.indexOf` 500 million times in a loop.

This is run on a P10 machine this time without `TR_DisableP10ArrayCmp` being set.
This compares my new implementation that doesn't have P10 optimizations with the old version that does have P10 optimizations.

The test being run is both the sub string and main string are the same length and have the same contents so there is a match at index 0.
This triggers the Arraycmp acceleration in the original implementation and the P10 version is used due to `TR_DisableP10ArrayCmp` not being set.

| Match Length | Old (s) | New (s) | Ratio |
| --- | --- | --- | --- |
| 1 | 1.14 | 0.46 | 0.40 |
| 2 | 2.01 | 0.66 | 0.33 |
| 3 | 2.30 | 0.65 | 0.28 |
| 4 | 2.30 | 0.95 | 0.41 |
| 5 | 2.29 | 0.90 | 0.39 |
| 6 | 2.30 | 1.02 | 0.45 |
| 7 | 2.29 | 1.02 | 0.45 |
| 8 | 2.30 | 1.02 | 0.45 |
| 9 | 2.29 | 1.02 | 0.45 |
| 10 | 2.29 | 1.27 | 0.56 |
| 11 | 2.29 | 1.27 | 0.56 |
| 12 | 2.30 | 1.40 | 0.61 |
| 13 | 2.29 | 1.40 | 0.61 |
| 14 | 2.29 | 1.53 | 0.67 |
| 15 | 2.30 | 1.52 | 0.66 |
| 16 | 2.29 | 1.15 | 0.50 |
| 17 | 2.52 | 1.15 | 0.45 |
| 18 | 2.53 | 1.40 | 0.55 |
| 19 | 2.52 | 1.40 | 0.55 |
| 20 | 2.53 | 1.65 | 0.65 |
| 21 | 2.53 | 1.65 | 0.65 |
| 22 | 2.53 | 1.78 | 0.70 |
| 23 | 2.53 | 1.77 | 0.70 |
| 24 | 2.53 | 1.65 | 0.65 |
| 25 | 2.53 | 1.65 | 0.65 |
| 26 | 2.53 | 1.90 | 0.75 |
| 27 | 2.53 | 1.90 | 0.75 |
| 28 | 2.52 | 2.02 | 0.80 |
| 29 | 2.52 | 2.02 | 0.80 |
| 30 | 2.52 | 2.15 | 0.85 |
| 31 | 2.53 | 2.15 | 0.85 |
| 32 | 2.53 | 1.40 | 0.55 |
| 33 | 2.76 | 1.40 | 0.51 |
| 34 | 2.76 | 1.65 | 0.60 |
| 35 | 2.76 | 1.65 | 0.60 |
| 36 | 2.76 | 1.90 | 0.69 |
| 37 | 2.76 | 1.90 | 0.69 |
| 38 | 2.76 | 2.03 | 0.73 |
| 39 | 2.76 | 2.03 | 0.73 |
| 40 | 2.76 | 1.91 | 0.69 |
| 41 | 2.77 | 1.90 | 0.69 |
| 42 | 2.76 | 2.16 | 0.78 |
| 43 | 2.76 | 2.16 | 0.78 |
| 44 | 2.76 | 2.28 | 0.83 |
| 45 | 2.76 | 2.28 | 0.83 |
| 46 | 2.77 | 2.41 | 0.87 |
| 47 | 2.76 | 2.43 | 0.88 |
| 48 | 2.76 | 1.68 | 0.61 |
| 49 | 2.99 | 1.68 | 0.56 |
| 50 | 2.99 | 1.90 | 0.64 |
| 51 | 2.99 | 1.90 | 0.64 |
| 52 | 2.99 | 2.15 | 0.72 |
| 53 | 2.99 | 2.15 | 0.72 |
| 54 | 2.99 | 2.28 | 0.76 |
| 55 | 2.99 | 2.28 | 0.76 |
| 56 | 2.99 | 2.15 | 0.72 |
| 57 | 2.99 | 2.15 | 0.72 |
| 58 | 2.99 | 2.41 | 0.81 |
| 59 | 2.99 | 2.41 | 0.81 |
| 60 | 2.99 | 2.53 | 0.85 |
| 61 | 3.00 | 2.53 | 0.84 |
| 62 | 2.99 | 2.66 | 0.89 |
| 63 | 2.99 | 2.66 | 0.89 |
| 64 | 2.99 | 2.15 | 0.72 |
| 128 | 4.90 | 4.13 | 0.84 |
| 256 | 8.29 | 7.60 | 0.92 |
| 512 | 11.27 | 11.65 | 1.03 |
| 1024 | 21.77 | 23.12 | 1.06 |

Performance is still better in almost all cases. The difference in performance for larger arrays is likely a testing quirk since the same 6 instruction loop is used for the bulk of the work performing the array cmp in both cases. So I think performance here is good enough to check in even without P10 optimization that can be added in the future.

IBMJimmyk commented:
Same as before, this test calls `StringLatin1.indexOf` 500 million times in a loop.

This time the numbers are for a Power 9 ppcle machine.

The first test being run is the full mismatch case. The first character of the substring can not be found anywhere in the main string.
This focuses on testing performance of searching for a match of the first character of the substring.

| Mismatch Length | Old (s) | New (s) | Ratio |
| --- | --- | --- | --- |
| 1 | 1.08 | 1.05 | 0.97 |
| 2 | 1.95 | 4.23 | 2.17 |
| 3 | 2.25 | 4.23 | 1.88 |
| 4 | 2.60 | 3.10 | 1.20 |
| 5 | 3.31 | 2.56 | 0.77 |
| 6 | 3.43 | 2.56 | 0.75 |
| 7 | 4.34 | 2.56 | 0.59 |
| 8 | 4.94 | 2.56 | 0.52 |
| 9 | 3.37 | 2.56 | 0.76 |
| 10 | 3.74 | 2.56 | 0.68 |
| 11 | 4.00 | 2.56 | 0.64 |
| 12 | 4.32 | 2.56 | 0.59 |
| 13 | 4.66 | 2.56 | 0.55 |
| 14 | 5.03 | 2.56 | 0.51 |
| 15 | 5.42 | 2.56 | 0.47 |
| 16 | 5.78 | 2.31 | 0.40 |
| 17 | 4.93 | 3.17 | 0.64 |
| 18 | 5.25 | 3.17 | 0.60 |
| 19 | 5.55 | 3.17 | 0.57 |
| 20 | 5.92 | 3.17 | 0.53 |
| 21 | 6.26 | 3.17 | 0.51 |
| 22 | 6.64 | 3.17 | 0.48 |
| 23 | 7.01 | 3.17 | 0.45 |
| 24 | 7.39 | 3.17 | 0.43 |
| 25 | 6.56 | 3.17 | 0.48 |
| 26 | 6.87 | 3.17 | 0.46 |
| 27 | 7.14 | 3.17 | 0.44 |
| 28 | 7.48 | 3.17 | 0.42 |
| 29 | 7.83 | 3.17 | 0.40 |
| 30 | 8.19 | 3.17 | 0.39 |
| 31 | 8.56 | 3.17 | 0.37 |
| 32 | 8.94 | 2.77 | 0.31 |
| 33 | 8.19 | 3.50 | 0.43 |
| 34 | 8.43 | 3.50 | 0.42 |
| 35 | 8.71 | 3.50 | 0.40 |
| 36 | 9.05 | 3.50 | 0.39 |
| 37 | 9.41 | 3.50 | 0.37 |
| 38 | 9.78 | 3.50 | 0.36 |
| 39 | 10.14 | 3.50 | 0.35 |
| 40 | 10.52 | 3.50 | 0.33 |
| 41 | 9.87 | 3.50 | 0.36 |
| 42 | 10.22 | 3.50 | 0.34 |
| 43 | 10.47 | 3.50 | 0.33 |
| 44 | 10.85 | 3.50 | 0.32 |
| 45 | 11.19 | 3.50 | 0.31 |
| 46 | 11.55 | 3.50 | 0.30 |
| 47 | 11.92 | 3.50 | 0.29 |
| 48 | 12.36 | 3.12 | 0.25 |
| 49 | 11.46 | 3.81 | 0.33 |
| 50 | 11.80 | 3.82 | 0.32 |
| 51 | 12.10 | 3.81 | 0.32 |
| 52 | 12.46 | 3.81 | 0.31 |
| 53 | 12.82 | 3.81 | 0.30 |
| 54 | 13.16 | 3.81 | 0.29 |
| 55 | 13.54 | 3.81 | 0.28 |
| 56 | 13.95 | 3.81 | 0.27 |
| 57 | 13.05 | 3.81 | 0.29 |
| 58 | 13.38 | 3.81 | 0.28 |
| 59 | 13.86 | 3.81 | 0.28 |
| 60 | 14.21 | 3.81 | 0.27 |
| 61 | 14.51 | 3.81 | 0.26 |
| 62 | 14.97 | 3.81 | 0.25 |
| 63 | 15.82 | 3.81 | 0.24 |
| 64 | 16.28 | 3.48 | 0.21 |
| 128 | 30.51 | 5.55 | 0.18 |
| 256 | 56.94 | 9.27 | 0.16 |
| 512 | 107.54 | 16.84 | 0.16 |
| 1024 | 208.59 | 35.25 | 0.17 |

My accelerated assembly has a special case to fast path a check of the first character of the substring with the first character of the main string so performance is as good as the original implementation which is naturally good with 1 byte long main strings. Performance isn't as good for 2, 3 and 4 byte long main string as byte by byte comparison still beats the vector load in performance. For 5 byte strings, the vector implementation is now better and becomes much better for longer strings.

This is once again the full match case where the sub string and main string have the same contents and will match at index 0. It is very similar to an arraycmp.

| Match Length | Old (s) | New (s) | Ratio |
| --- | --- | --- | --- |
| 1 | 1.20 | 0.51 | 0.43 |
| 2 | 2.42 | 0.80 | 0.33 |
| 3 | 2.98 | 0.80 | 0.27 |
| 4 | 3.27 | 0.98 | 0.30 |
| 5 | 2.65 | 0.98 | 0.37 |
| 6 | 2.98 | 1.11 | 0.37 |
| 7 | 3.31 | 1.11 | 0.33 |
| 8 | 3.61 | 1.10 | 0.31 |
| 9 | 2.87 | 1.13 | 0.39 |
| 10 | 3.29 | 1.36 | 0.41 |
| 11 | 3.62 | 1.41 | 0.39 |
| 12 | 3.90 | 1.50 | 0.39 |
| 13 | 3.29 | 1.54 | 0.47 |
| 14 | 3.62 | 1.64 | 0.45 |
| 15 | 3.92 | 1.69 | 0.43 |
| 16 | 4.19 | 1.64 | 0.39 |
| 17 | 3.34 | 1.64 | 0.49 |
| 18 | 3.76 | 1.84 | 0.49 |
| 19 | 4.03 | 1.91 | 0.47 |
| 20 | 4.28 | 2.12 | 0.49 |
| 21 | 3.85 | 2.19 | 0.57 |
| 22 | 4.11 | 2.25 | 0.55 |
| 23 | 4.39 | 2.33 | 0.53 |
| 24 | 4.62 | 2.12 | 0.46 |
| 25 | 3.92 | 2.19 | 0.56 |
| 26 | 4.28 | 2.40 | 0.56 |
| 27 | 4.47 | 2.47 | 0.55 |
| 28 | 4.73 | 2.53 | 0.54 |
| 29 | 4.28 | 2.61 | 0.61 |
| 30 | 4.48 | 2.67 | 0.60 |
| 31 | 4.77 | 2.75 | 0.58 |
| 32 | 5.00 | 2.12 | 0.42 |
| 33 | 3.86 | 2.19 | 0.57 |
| 34 | 4.29 | 2.33 | 0.54 |
| 35 | 4.57 | 2.49 | 0.54 |
| 36 | 4.79 | 2.62 | 0.55 |
| 37 | 4.39 | 2.83 | 0.64 |
| 38 | 4.65 | 2.74 | 0.59 |
| 39 | 4.92 | 2.98 | 0.61 |
| 40 | 5.13 | 2.62 | 0.51 |
| 41 | 4.40 | 2.85 | 0.65 |
| 42 | 4.76 | 3.00 | 0.63 |
| 43 | 5.01 | 3.12 | 0.62 |
| 44 | 5.25 | 3.16 | 0.60 |
| 45 | 4.76 | 3.28 | 0.69 |
| 46 | 5.01 | 3.34 | 0.67 |
| 47 | 5.31 | 3.44 | 0.65 |
| 48 | 5.55 | 2.66 | 0.48 |
| 49 | 4.34 | 2.77 | 0.64 |
| 50 | 4.72 | 2.90 | 0.61 |
| 51 | 5.00 | 3.16 | 0.63 |
| 52 | 5.22 | 3.12 | 0.60 |
| 53 | 4.81 | 3.44 | 0.72 |
| 54 | 5.08 | 3.26 | 0.64 |
| 55 | 5.35 | 3.57 | 0.67 |
| 56 | 5.57 | 3.12 | 0.56 |
| 57 | 4.81 | 3.43 | 0.71 |
| 58 | 5.17 | 3.53 | 0.68 |
| 59 | 5.44 | 3.69 | 0.68 |
| 60 | 5.68 | 3.71 | 0.65 |
| 61 | 5.17 | 3.86 | 0.75 |
| 62 | 5.44 | 3.92 | 0.72 |
| 63 | 5.74 | 4.06 | 0.71 |
| 64 | 5.99 | 3.24 | 0.54 |
| 128 | 7.73 | 5.74 | 0.74 |
| 256 | 11.28 | 10.99 | 0.97 |
| 512 | 18.48 | 21.46 | 1.16 |
| 1024 | 34.97 | 43.88 | 1.25 |

Once again, performance is good for shorter strings but seemingly not quite as good for longer strings. Once again, I think this is due to a testing quirk since both versions will run the same instructions as a vector loop to perform most of the arraycmp for longer arrays.

IBMJimmyk commented:
These results are for the same tests but for AIX on P9. Overall they are similar but slightly better than the ppcle P9 results.
Only the 2 and 3 length mismatched strings lengths are slower and not by as much.
For matching strings, performance is always better.

| Mismatch Length | Old (s) | New (s) | Ratio |
| --- | --- | --- | --- |
| 1 | 1.08 | 0.66 | 0.61 |
| 2 | 1.96 | 2.56 | 1.31 |
| 3 | 2.26 | 2.56 | 1.13 |
| 4 | 2.60 | 2.56 | 0.98 |
| 5 | 3.29 | 2.56 | 0.78 |
| 6 | 3.44 | 2.56 | 0.75 |
| 7 | 4.35 | 2.56 | 0.59 |
| 8 | 4.95 | 2.56 | 0.52 |
| 9 | 3.38 | 2.56 | 0.76 |
| 10 | 3.75 | 2.56 | 0.68 |
| 11 | 4.00 | 2.56 | 0.64 |
| 12 | 4.33 | 2.56 | 0.59 |
| 13 | 4.67 | 2.56 | 0.55 |
| 14 | 5.03 | 2.56 | 0.51 |
| 15 | 5.43 | 2.56 | 0.47 |
| 16 | 5.80 | 2.31 | 0.40 |
| 17 | 4.94 | 3.18 | 0.64 |
| 18 | 5.25 | 3.17 | 0.60 |
| 19 | 5.54 | 3.17 | 0.57 |
| 20 | 5.93 | 3.17 | 0.54 |
| 21 | 6.28 | 3.17 | 0.51 |
| 22 | 6.65 | 3.17 | 0.48 |
| 23 | 7.02 | 3.17 | 0.45 |
| 24 | 7.40 | 3.17 | 0.43 |
| 25 | 6.56 | 3.17 | 0.48 |
| 26 | 6.88 | 3.17 | 0.46 |
| 27 | 7.15 | 3.17 | 0.44 |
| 28 | 7.49 | 3.17 | 0.42 |
| 29 | 7.84 | 3.17 | 0.40 |
| 30 | 8.20 | 3.17 | 0.39 |
| 31 | 8.57 | 3.17 | 0.37 |
| 32 | 8.95 | 2.78 | 0.31 |
| 33 | 8.15 | 3.51 | 0.43 |
| 34 | 8.44 | 3.51 | 0.42 |
| 35 | 8.72 | 3.51 | 0.40 |
| 36 | 9.06 | 3.51 | 0.39 |
| 37 | 9.41 | 3.51 | 0.37 |
| 38 | 9.79 | 3.51 | 0.36 |
| 39 | 10.15 | 3.51 | 0.35 |
| 40 | 10.53 | 3.51 | 0.33 |
| 41 | 9.88 | 3.51 | 0.36 |
| 42 | 10.22 | 3.51 | 0.34 |
| 43 | 10.48 | 3.51 | 0.33 |
| 44 | 10.87 | 3.51 | 0.32 |
| 45 | 11.19 | 3.51 | 0.31 |
| 46 | 11.56 | 3.51 | 0.30 |
| 47 | 11.94 | 3.51 | 0.29 |
| 48 | 12.37 | 3.12 | 0.25 |
| 49 | 11.46 | 3.82 | 0.33 |
| 50 | 11.78 | 3.82 | 0.32 |
| 51 | 12.14 | 3.82 | 0.31 |
| 52 | 12.44 | 3.82 | 0.31 |
| 53 | 12.83 | 3.82 | 0.30 |
| 54 | 14.76 | 3.82 | 0.26 |
| 55 | 13.53 | 3.82 | 0.28 |
| 56 | 13.95 | 3.82 | 0.27 |
| 57 | 13.05 | 3.82 | 0.29 |
| 58 | 13.38 | 3.82 | 0.29 |
| 59 | 13.96 | 3.82 | 0.27 |
| 60 | 14.11 | 3.82 | 0.27 |
| 61 | 14.47 | 3.82 | 0.26 |
| 62 | 14.81 | 3.82 | 0.26 |
| 63 | 15.29 | 3.82 | 0.25 |
| 64 | 15.84 | 3.49 | 0.22 |
| 128 | 30.46 | 5.55 | 0.18 |
| 256 | 56.99 | 9.22 | 0.16 |
| 512 | 107.58 | 16.79 | 0.16 |
| 1024 | 210.30 | 33.27 | 0.16 |

| Match Length | Old (s) | New (s) | Ratio |
| --- | --- | --- | --- |
| 1 | 1.20 | 0.64 | 0.53 |
| 2 | 2.28 | 0.98 | 0.43 |
| 3 | 2.99 | 0.97 | 0.33 |
| 4 | 3.27 | 0.99 | 0.30 |
| 5 | 2.65 | 0.99 | 0.37 |
| 6 | 2.99 | 1.11 | 0.37 |
| 7 | 3.30 | 1.11 | 0.34 |
| 8 | 3.62 | 1.11 | 0.31 |
| 9 | 2.88 | 1.14 | 0.40 |
| 10 | 3.30 | 1.50 | 0.46 |
| 11 | 3.62 | 1.51 | 0.42 |
| 12 | 3.91 | 1.51 | 0.39 |
| 13 | 3.27 | 1.54 | 0.47 |
| 14 | 3.62 | 1.64 | 0.45 |
| 15 | 3.92 | 1.70 | 0.43 |
| 16 | 4.19 | 1.48 | 0.35 |
| 17 | 3.38 | 1.45 | 0.43 |
| 18 | 3.77 | 1.70 | 0.45 |
| 19 | 4.04 | 1.72 | 0.42 |
| 20 | 4.29 | 1.95 | 0.45 |
| 21 | 3.83 | 1.98 | 0.52 |
| 22 | 4.15 | 2.08 | 0.50 |
| 23 | 4.39 | 2.11 | 0.48 |
| 24 | 4.62 | 1.95 | 0.42 |
| 25 | 3.93 | 1.99 | 0.51 |
| 26 | 4.28 | 2.22 | 0.52 |
| 27 | 4.50 | 2.26 | 0.50 |
| 28 | 4.76 | 2.34 | 0.49 |
| 29 | 4.28 | 2.39 | 0.56 |
| 30 | 4.50 | 2.47 | 0.55 |
| 31 | 4.78 | 2.52 | 0.53 |
| 32 | 5.01 | 1.85 | 0.37 |
| 33 | 3.87 | 1.88 | 0.49 |
| 34 | 4.30 | 2.05 | 0.48 |
| 35 | 4.57 | 2.18 | 0.48 |
| 36 | 4.79 | 2.33 | 0.49 |
| 37 | 4.40 | 2.47 | 0.56 |
| 38 | 4.67 | 2.44 | 0.52 |
| 39 | 4.92 | 2.64 | 0.54 |
| 40 | 5.14 | 2.32 | 0.45 |
| 41 | 4.40 | 2.49 | 0.57 |
| 42 | 4.77 | 2.67 | 0.56 |
| 43 | 5.02 | 2.77 | 0.55 |
| 44 | 5.26 | 2.81 | 0.53 |
| 45 | 4.77 | 2.91 | 0.61 |
| 46 | 5.02 | 2.94 | 0.59 |
| 47 | 5.31 | 3.03 | 0.57 |
| 48 | 5.55 | 2.25 | 0.41 |
| 49 | 4.35 | 2.24 | 0.51 |
| 50 | 4.74 | 2.42 | 0.51 |
| 51 | 5.01 | 2.53 | 0.50 |
| 52 | 5.22 | 2.84 | 0.54 |
| 53 | 4.82 | 2.97 | 0.62 |
| 54 | 5.10 | 2.94 | 0.58 |
| 55 | 5.36 | 3.15 | 0.59 |
| 56 | 5.58 | 2.70 | 0.48 |
| 57 | 4.82 | 2.89 | 0.60 |
| 58 | 5.17 | 3.09 | 0.60 |
| 59 | 5.46 | 3.17 | 0.58 |
| 60 | 5.68 | 3.21 | 0.57 |
| 61 | 5.17 | 3.29 | 0.64 |
| 62 | 5.46 | 3.35 | 0.61 |
| 63 | 5.75 | 3.42 | 0.59 |
| 64 | 6.00 | 2.64 | 0.44 |
| 128 | 7.74 | 4.22 | 0.55 |
| 256 | 11.30 | 7.53 | 0.67 |
| 512 | 18.49 | 14.63 | 0.79 |
| 1024 | 35.31 | 29.29 | 0.83 |

IBMJimmyk commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -12418,6 +12418,905 @@ static TR::Register *inlineIntrinsicIndexOf(TR::Node *node, TR::CodeGenerator *c
    }
 
+static void performArrayCmp(TR::Node *node, TR::CodeGenerator *cg, TR_PPCScratchRegisterManager *srm, TR::LabelSymbol *subStringMatchLabel, TR::LabelSymbol *subStringMismatchLabel, TR::Register *currentAddressReg, TR::Register *subStringAddressReg, TR::Register *subStringLenReg, TR::Register *vec1Reg, TR::Register *vec2Reg, TR::Register *tempReg, TR::Register *cr0Reg, TR::Register *cr6Reg)
+   {
```
Yes, P10 will be able to do better here but I that implementation is currently future work.

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -12418,6 +12418,905 @@ static TR::Register *inlineIntrinsicIndexOf(TR::Node *node, TR::CodeGenerator *c
+      /* If offset + number of bytes in the substring is greater than the number of bytes in the main string it can't be found. In this case jump to notFoundLabel to return -1. */
+      TR::Register *minStrByteCountReg = srm->findOrCreateScratchRegister();
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, minStrByteCountReg, fromIndexReg, strCountReg);
+      if (!isLatin1)
```
isn't it weird that arrayLen is represented in terms of byte, while fromIndex and strCount are expected to be represented in terms of element?

if all are consistently meaning element, you don't need to do this doubling, that is.

IBMJimmyk replied:
The line right after this was an error. arrayLen, strCount and fromIndex all contain a element count or index to begin with. So the value does not need to be adjusted even under UTF16. I've pushed a change to fix this.

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -12418,6 +12418,905 @@ static TR::Register *inlineIntrinsicIndexOf(TR::Node *node, TR::CodeGenerator *c
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, currentAddressReg, arrayAddressReg, resultReg);
+
+   TR::Register *firstCharReg = srm->findOrCreateScratchRegister();
+   TR::Register *firstCharSubStrReg = srm->findOrCreateScratchRegister();
```
i am wondering if you can optimize the use of these many registers. at this point, 13 or 14 GPRs are needed already.

IBMJimmyk replied:
My current implementation uses:
11-16 GPR (Depends on how many input Regs have ref count of 0 and can be reused. Saves up to 5 registers)
7 VRF
2 CCR

For comparison, the Arm implementation:
https://github.com/eclipse-openj9/openj9/blob/45733dc106ac9a52570239011e90eb61ae01fd74/runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp#L6672-L6701
This uses:
11-13 GPR (Depends on how many input Regs have ref count of 0 and can be reused. Saves up to 2 registers)
3 VRF

IBMJimmyk commented:
It took a bit longer than I had hoped, but I reworked the assembly to reduce the number of registers being used.

It now uses:
9-13 GPR (Depends on how many input Regs have ref count of 0 and can be reused. Saves up to 4 registers)
7 VRF
2 CCR

GPR count before the recent changes was 11-16 GPR. This is a reduction of 2-3 registers.

Early tests are passing so things currently seem functional. I still have more tests to run and performance tests as well.

IBMJimmyk commented:
The performance hit from reducing the number of registers turned out to be larger than expected so the expected route is to go back to the version that uses more registers even if it might mean more register pressure.

But, I wanted to make one final attempt at trying to fix the version that uses fewer registers or at least understand what might have gone wrong for future reference.

The test I has two string.
The string being searched in is the letter "a" repeated 1025 times.
The string being searched for is "ab".
This case is meant to exercise the code path for handling the first character of the sub string being found but not the rest of the string.

For the high register version:
```
perf stat -e cycles,instructions java -Xmx8g -Xms8g -Xjit:"disableGLU,trytoinline={*},accepthugemethods,limit={*runTest*},optlevel=hot,count=5,disableasynccompilation" -Xshareclasses:none -Xnoaot -XX:-EnableHCR -DarraySize=1025 -DsubArraySize=2 -DmatchIndex1=-1 -DpartialIndex1=-1 -DpartialIndex2=-1 -DmismatchIndex1=-1 -DsearchOffset=0 -Diterations=5000 -DpartialBase=1 IndexOfStrPerfSimple
Options: arraySize=1025, subArraySize=2, matchIndex1=-1, partialIndex1=-1, partialIndex2=-1, mismatchIndex1=-1, searchOffset=0, iterations=5000, partialBase=1
Warmup Time(s): 10.705
Result: -10 (-10000)
Measured Time(s): 261.442

 Performance counter stats for 'java -Xmx8g -Xms8g -Xjit:disableGLU,trytoinline={*},accepthugemethods,limit={*runTest*},optlevel=hot,count=5,disableasynccompilation -Xshareclasses:none -Xnoaot -XX:-EnableHCR -DarraySize=1025 -DsubArraySize=2 -DmatchIndex1=-1 -DpartialIndex1=-1 -DpartialIndex2=-1 -DmismatchIndex1=-1 -DsearchOffset=0 -Diterations=5000 -DpartialBase=1 IndexOfStrPerfSimple':

 1,089,215,878,710      cycles
 1,479,290,355,209      instructions                     #    1.36  insn per cycle

     272.409532252 seconds time elapsed

     272.477779000 seconds user
       0.196021000 seconds sys
```
And the lower register version:
```
TR_EnableStringIndexOfStringIntrinsicV2=1 perf stat -e cycles,instructions java -Xmx8g -Xms8g -Xjit:"disableGLU,trytoinline={*},accepthugemethods,limit={*runTest*},optlevel=hot,count=5,disableasynccompilation" -Xshareclasses:none -Xnoaot -XX:-EnableHCR -DarraySize=1025 -DsubArraySize=2 -DmatchIndex1=-1 -DpartialIndex1=-1 -DpartialIndex2=-1 -DmismatchIndex1=-1 -DsearchOffset=0 -Diterations=5000 -DpartialBase=1 IndexOfStrPerfSimple
Options: arraySize=1025, subArraySize=2, matchIndex1=-1, partialIndex1=-1, partialIndex2=-1, mismatchIndex1=-1, searchOffset=0, iterations=5000, partialBase=1
JIT: env var TR_EnableStringIndexOfStringIntrinsicV2 is set to 1
Warmup Time(s): 8.812
Result: -10 (-10000)
Measured Time(s): 334.590

 Performance counter stats for 'java -Xmx8g -Xms8g -Xjit:disableGLU,trytoinline={*},accepthugemethods,limit={*runTest*},optlevel=hot,count=5,disableasynccompilation -Xshareclasses:none -Xnoaot -XX:-EnableHCR -DarraySize=1025 -DsubArraySize=2 -DmatchIndex1=-1 -DpartialIndex1=-1 -DpartialIndex2=-1 -DmismatchIndex1=-1 -DsearchOffset=0 -Diterations=5000 -DpartialBase=1 IndexOfStrPerfSimple':

 1,374,105,939,819      cycles
 1,355,779,076,243      instructions                     #    0.99  insn per cycle

     343.677190478 seconds time elapsed

     343.804909000 seconds user
       0.180017000 seconds sys
```
The number of instructions execute goes down a bit which is supposed to be a positive but the IPC falls from 1.36 to 0.99 which causes performance to tank instead.

Annotated assembly of the hot section looks like this.
High register version (27 instruction loop, 97% of the total ticks):
```
       4 :   4bc:    lvsl    v12,r7,r7
       0 :   4c0:    vspltisb v7,15
      12 :   4c4:    vsububm v12,v7,v12
       0 :   4c8:    vspltisb v7,3
   45665 :   4cc:    vslb    v12,v12,v7
      95 :   4d0:    vbpermq v7,v8,v12
  117010 :   4d4:    mfvrwz  r3,v7
      63 :   4d8:    cntlzw  r3,r3
   69614 :   4dc:    addi    r3,r3,-16
       0 :   4e0:    add     r9,r3,r9
       0 :   4e4:    clrrdi  r6,r0,1
       0 :   4e8:    cmpldi  cr6,r6,2
       0 :   4ec:    blt     cr6,5b0 <LIndexOfStrPerfSimple;runTest([B[BI)I+0x530>
       0 :   4f0:    andi.   r2,r0,1
       0 :   4f4:    cmpldi  cr6,r6,4
       0 :   4f8:    blt     cr6,584 <LIndexOfStrPerfSimple;runTest([B[BI)I+0x504>  //<<--always taken

   94053 :   584:    lhzx    r6,r2,r9
       1 :   588:    lhzx    r4,r2,r10
   71089 :   58c:    cmpw    cr6,r6,r4
       0 :   590:    beq     cr6,5b0 <LIndexOfStrPerfSimple;runTest([B[BI)I+0x530>
       0 :   594:    subf    r9,r3,r9
       0 :   598:    lvsl    v12,r7,r3
       0 :   59c:    vperm   v7,v13,v13,v12
   44361 :   5a0:    vsel    v11,v11,v10,v7
       0 :   5a4:    vcmpequb. v8,v11,v9
   22736 :   5a8:    beq     cr6,2d8 <LIndexOfStrPerfSimple;runTest([B[BI)I+0x258>
      10 :   5ac:    b       4bc <LIndexOfStrPerfSimple;runTest([B[BI)I+0x43c>
```
Lower register version (25 instruction loop, 98% of the total ticks):
```
    3535 :   368:    vbpermq v8,v9,v13
   68130 :   36c:    mfvrwz  r3,v8
    2017 :   370:    slwi    r3,r3,16
   50082 :   374:    andi.   r6,r4,15
    4614 :   378:    slw.    r3,r3,r6
   51301 :   37c:    beq     1b0 <LIndexOfStrPerfSimple;runTest([B[BI)I+0x130>
    2083 :   380:    cntlzw  r3,r3
   53354 :   384:    add     r4,r3,r4
       0 :   388:    cmpd    r4,r5
   18772 :   38c:    bge     1bc <LIndexOfStrPerfSimple;runTest([B[BI)I+0x13c>
       0 :   390:    clrrdi  r3,r7,1
       0 :   394:    cmpldi  cr6,r3,2
       2 :   398:    blt     cr6,45c <LIndexOfStrPerfSimple;runTest([B[BI)I+0x3dc>
       0 :   39c:    andi.   r6,r7,1
       0 :   3a0:    cmpldi  cr6,r3,4
       0 :   3a4:    blt     cr6,430 <LIndexOfStrPerfSimple;runTest([B[BI)I+0x3b0>  //<<--always taken

   38260 :   430:    lhzx    r3,r6,r4
       0 :   434:    lhzx    r8,r6,r2
   59465 :   438:    cmpw    cr6,r3,r8
       0 :   43c:    beq     cr6,45c <LIndexOfStrPerfSimple;runTest([B[BI)I+0x3dc>
       0 :   440:    lvsl    v9,0,r4
       0 :   444:    vperm   v8,v14,v15,v9
   65076 :   448:    vsel    v12,v11,v12,v8
     262 :   44c:    vcmpequb. v9,v12,v10
   51891 :   450:    bne     cr6,368 <LIndexOfStrPerfSimple;runTest([B[BI)I+0x2e8>
```
This section of the assembly handles the first character match followed by not finding the rest of the string. Just before this section of code, a vector register is loaded with 16 characters from the main string. This will always be "aaaaaaaaaaaaaaaa" (16 a) for this test case. This loop is then run 16 times before going back to load another 16 bytes. And this is repeated for the entire 1025 byte String before returning a result that the "ab" string was not found.

Both versions have the same number of taken branches and the lower register version runs fewer instructions. But, the IPC is a lot worse so it performs worse.

@zl-wang Do you have any ideas on why this might be the case?

zl-wang commented:
> @zl-wang Do you have any ideas on why this might be the case?

a quick casual reading indicated the slow-down can be mostly attributed to the top/earlier section (~10% slower ... assuming roughly same number of iterations over it). and, as far as i can see, the reason for slowing down is mainly for more sequential WR dependency of registers, not much pipeline hazard. given limited number of registers, not sure if you can break down the dependency.  let's get the better-performing version in ... move on, especially given the benefit in real workload, instead of this zoom-in scenario.  we may come back to see if we can improve further on the slower version at lower priority.

IBMJimmyk commented:
Okay, I will focus on the better performing higher register version and maybe this can be revisited at a later date.

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -12358,6 +12358,936 @@ static TR::Register *inlineIntrinsicIndexOf(TR::Node *node, TR::CodeGenerator *c
+   /* If a match is found, jump to subStringMatchLabel to handle the substring match. */
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::cmp4, node, cr6Reg, tempReg, temp2Reg);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::beq, node, subStringMatchLabel, cr6Reg);
+   /* A fallthrough means the loaded values did not match. Continue on to handle the case where the substring was not found. */
```
i assumed the case where main string is shorter than sub-string is handled/guarded outside this routine.

IBMJimmyk replied:
That is correct. It is handled in two places:

https://github.com/eclipse-openj9/openj9/blob/75f0b5ee0c00bfb4dde15664ba27beefa0d089b6/runtime/compiler/p/codegen/J9TreeEvaluator.cpp#L12646-L12649

https://github.com/eclipse-openj9/openj9/blob/75f0b5ee0c00bfb4dde15664ba27beefa0d089b6/runtime/compiler/p/codegen/J9TreeEvaluator.cpp#L12720-L12737

After the second section `endAddressReg` contains the address where the sub string is now too long to be found in the rest of the main string. As a result, it won't go to the code generated by `performArrayCmp` since a match is impossible.

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -12358,6 +12358,936 @@ static TR::Register *inlineIntrinsicIndexOf(TR::Node *node, TR::CodeGenerator *c
+    * This is done so one fewer byte is compared in odd cases.
+    * tempReg holds the actual number of bytes that need to be compared and is always an even number.
+    * If tempReg is less than 2, it must be 0 so a full substring match has been found.
+    */
```
observation of inconsistency:  offset, length, and index values seemed being treated as 64bit in this area of code, but not in others. preference: treating them consistently.

IBMJimmyk replied:
I can try looking over this.

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -12358,6 +12358,936 @@ static TR::Register *inlineIntrinsicIndexOf(TR::Node *node, TR::CodeGenerator *c
+    */
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::subf, node, endAddressReg, subStringLenReg, endAddressReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::cmp8, node, cr0Reg, resultReg, endAddressReg);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::bge, node, notFoundLabel, cr0Reg);
```
compare-equal is included in this branch ... due to NotFoundEarly and before offset-reg (resultReg) is adjusted. walking on a pretty thin line ... way of programming.

IBMJimmyk replied:
I also thought this was very awkward. But, doing it this way let me avoid incrementing `endAddressReg` if the branch is taken. I made sure to explicitly test this area to make sure the math is right.

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -12358,6 +12358,936 @@ static TR::Register *inlineIntrinsicIndexOf(TR::Node *node, TR::CodeGenerator *c
+    * It sometimes isn't necessary to match the first character of the substring against the last characters in the main string
+    * because there aren't enough characters left in the main string to hold the rest of the substring.
+    * A check for this case is performed here and goes to notFoundLabel to return -1 if there aren't enough characters left.
+    */
```
can strCount (or subStringLen) be zero? i.e. do you need to handle that case?

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -12358,6 +12358,936 @@ static TR::Register *inlineIntrinsicIndexOf(TR::Node *node, TR::CodeGenerator *c
+   /*
+    * Calculate the actual addresses of the start and end points.
+    * currentAddressReg is the address of the next character to load in the main string to compare against the substring.
+    * endAddressReg is the address of the first character in the main string that does NOT need to be compared against the first character of the substring.
```
does NOT need to ...  a better comment to me is:  we will end the search at this address, because we already know there is no match at this point.

IBMJimmyk commented on runtime/compiler/p/codegen/J9CodeGenerator.cpp:
```diff
@@ -98,6 +98,13 @@ J9::Power::CodeGenerator::initialize()
+      comp->target().is64Bit() && !comp->getOption(TR_DisableFastStringIndexOf) &&
+      !TR::Compiler->om.canGenerateArraylets() &&
+      !disableIndexOfInlining)
+      cg->setSupportsInlineStringIndexOfString();
```
I will rename it to `TR_DisableStringIndexOfStringInlining` since it guards the call to `setSupportsInlineStringIndexOfString`. The existing `TR_DisableFastStringIndexOf` options disabled more than just this flag and I wanted to be able to turn only my changes on or off.

From this, I actually found a bug where `StringUTF16.indexOf` won't get inlined if `setSupportsInlineStringIndexOfString` is set since it is expecting the call to be replaced with inlined assembly. This doesn't happen yet so I will change it to continue with normal inlining of the UTF16 methods for now.

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -12358,6 +12358,936 @@ static TR::Register *inlineIntrinsicIndexOf(TR::Node *node, TR::CodeGenerator *c
+   /*
+    * Now our vector is ready for comparison: no garbage can match the target value and the start of
+    * the vector is now aligned to the start of the string.
+    * This jumps to foundFirstVectorLabel to check the rest of the substring. If it doesn't match, it jumps back to firstContinueLabel.
```
if it doesn't match, it falls through to firstContinueLabel, instead of jumping back to.

IBMJimmyk replied:
My intent for this was to say if the branch was taken and the rest of the substring isn't found, it will jump back to `firstContinueLabel`. I edited the comment to be more clear.

IBMJimmyk commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -12358,6 +12358,936 @@ static TR::Register *inlineIntrinsicIndexOf(TR::Node *node, TR::CodeGenerator *c
+      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::vsplth, node, targetVectorReg, targetVectorReg, 3);
+
+   /*
+    * targetVectorNotReg is a vector register set to all 0s.
```
I'm changing this line to "targetVectorNotReg is a vector register temporarily set to all 0s.".
`targetVectorNotReg` is used as a temporary register here to set up the constant in `maskVectorReg`.
A few instructions later `targetVectorNotReg` is set the bitwise complement of `targetVectorReg` which is the value it holds for the remainder of the `indexOf` operation.

IBMJimmyk commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -12358,6 +12358,936 @@ static TR::Register *inlineIntrinsicIndexOf(TR::Node *node, TR::CodeGenerator *c
+
+   /* This checks to see if there are any characters left to look at. */
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::cmp8, node, cr0Reg, currentAddressReg, endAddressReg);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::beq, node, notFoundLabel, cr0Reg);
```
`currentAddressReg` will never be greater than `endAddressReg`. If it is, something is already broken. Right now, this error condition will lead into an infinite loop and read off the end of the array until a segfault.  Changing this to `bge` would mean returning -1 instead. That might be a bit safer so I'll make the change. But, once again this only happens if there's an error.

IBMJimmyk commented:
I still have a few items in the review to address but I've pushed a commit with my latest changes. Changes include:

StringUTF16 methods that was erroneously prevented from inlining are now eligible for inlining again
Renamed `TR_DisableIndexOfInlining` to `TR_DisableStringIndexOfStringInlining`
Changed 32bit compares to 64 bit compares for consistency
Improved wording in several comments.
Changed compares from `beq` to `bge` to better handle unexpected errors.

Made a simple code sequence change:
Before:
```
  addi      resultReg, resultReg, 1     //add 2 for UTF16
  addi      endAddressReg, endAddressReg, 1       //add 2 for UTF16
  add       currentAddressReg, arrayAddressReg, resultReg
```
After:
```
  addi      currentAddressReg, currentAddressReg, 1     //add 2 for UTF16
  addi      endAddressReg, endAddressReg, 1       //add 2 for UTF16
```

IBMJimmyk commented:
The value in `resultReg` is not needed anymore at that point. It gets used to hold temporary values from then on. At the very end, it gets filled in with the correct return value.

IBMJimmyk commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -12358,6 +12358,936 @@ static TR::Register *inlineIntrinsicIndexOf(TR::Node *node, TR::CodeGenerator *c
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::subf, node, resultReg, arrayAddressReg, currentAddressReg);
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, foundExactLabel);
+
```
Do you have any specific suggestions about the duplicated code? By number of instructions, do you mean the total number of instructions generated? I have not checked that against Arm.

zl-wang replied:
as it is, it is going to be tough in reusing (by jumping to) a section of code, because different labelSymbol(s) are passed into `performArrayCmp`. if you can re-organize the code such that, each time `performArrayCmp` is called, two same LabelSymbol(s) are passed in, then each later call of `performArrayCmp` can be changed into a branch to the starting of the first call.

this principle is applicable to the compare and search repetitions of codegen too.

i would say this is less priority than the goal to merge this in the next few days. Just take a look and see if this is doable easily.

IBMJimmyk commented:
I think I've addressed all comments so far. I've also reran some quick functional tests to make sure I didn't break anything unexpectedly.

zl-wang commented:
Jenkins test sanity aix,plinux jdk8,jdk21
